mitosheet/mitosheet/step_performers/import_steps/snowflake_import.py

```diff
@@ -0,0 +1,149 @@
+        # TODO: We don't want to send these credentials because then they get saved to the analysis!!
+        # Instead, we can save them in the api call and read them in here. That is a weird flow because it 
+        # means the step has dependencies on the api call and the order of operations in a way we've never had before. 
+        # When the user reruns the analysis, if there is a snowflake_import, we need to have them reautnethicate first (?)
+        credentials: SnowflakeCredentials = get_param(params, 'credentials') 
+        table_loc_and_warehouse: SnowflakeTableLocationAndWarehouse = get_param(params, 'table_loc_and_warehouse')
+        query_params: SnowflakeQueryParams = get_param(params, 'query_params')
+
+        # We make a new state to modify it
+        post_state = prev_state.copy()
+
+        pandas_start_time = perf_counter()
+        table = table_loc_and_warehouse['table']
+
+        connection_params_dict = get_connection_param_dict(credentials, table_loc_and_warehouse)
+        
+        try: 
+            # First try to establish the connection
+            con = snowflake.connector.connect(**connection_params_dict)
+        except Exception as e: 
+            print(e)
+            # When we do the frontend, we can figure out exactly what we want to raise here
+            raise make_invalid_snowflake_import_error()
+        
+        try:
+            # Second execute the query
+            cur = con.cursor()
+            sql_query = create_query(table, query_params)
+            cur.execute(sql_query)
+            df = cur.fetch_pandas_all()
+        except: 
+            raise make_invalid_snowflake_import_error()
+        finally:
+           # If we've created the connection, then make sure to close it
+           con.close() # type: ignore
+
+        new_df_name = get_valid_dataframe_name(post_state.df_names, table)
+        post_state.add_df_to_state(
+            df, 
+            DATAFRAME_SOURCE_IMPORTED, 
+            df_name=new_df_name,
+        )   
+
+        pandas_processing_time = perf_counter() - pandas_start_time
+
+        return post_state, {
+            'pandas_processing_time': pandas_processing_time,
+            'result': {
+                # TODO: fill in the result
+            }
+        }
+
+    @classmethod
+    def transpile(
+        cls,
+        prev_state: State,
+        post_state: State,
+        params: Dict[str, Any],
+        execution_data: Optional[Dict[str, Any]],
+    ) -> List[CodeChunk]:
+
+        table_loc_and_warehouse: SnowflakeTableLocationAndWarehouse = get_param(params, 'table_loc_and_warehouse')
+        connection_params_dict = get_connection_param_dict(get_param(params, 'credentials'), table_loc_and_warehouse)
+
+        query_params: SnowflakeQueryParams = get_param(params, 'query_params')
+        table = table_loc_and_warehouse['table']
+        sql_query = create_query(table, query_params)
```

Comment:
There's one final refactor here. You should:
1. Return these variables (namely, `connection_params_dict` and `sql_query`) in execution data above.
2. Get it form the execution data in this function.

This means you have to redo _zero_ work -- which again reduces the chances we'll make silly bugs and forget to change things.